### PR TITLE
Raylib Driverstation Current Progress

### PR DESCRIPTION
**Pull request by PhocaeaUnitas**
See issue #8
See issue #17

### CODE REVIEW

File: src/driverstation/.gitignore
Comment (yameatmeyourdead) [inline src/driverstation/.gitignore:1]: Why in this commit :)

File: src/driverstation/build.sh
Comment (yameatmeyourdead) [inline src/driverstation/build.sh:1]: git ignore this

File: src/driverstation/run.sh
Comment (yameatmeyourdead) [inline src/driverstation/run.sh:1]: git ignore this

File: src/driverstation/include/ssh_ui.hpp [hunk 1]
```diff
@@ -0,0 +1,771 @@
...
+							default:
+								std::cout << "Unhandled SGR parameter: " << parameters[i].value << std::endl;
+								break;
+						}
+					}
+					break;
+				default:
+					// TODO: Temporary
+					std::cout << "Unknown escape code: \\x1B[";
+					for(int i = 0; i < this->current_csi_parameters.size(); i++){
```
Comment (yameatmeyourdead) [inline src/driverstation/include/ssh_ui.hpp:710]: cast i to size_t to prevent compiler warning about signed/unsigned comparison

**Review by yameatmeyourdead — COMMENTED** [review]
Pretty good so far. Most of my current requests for changes deal with compiler warnings. Try developing with `-Wall -Wextra -Wpendantic` to catch these. Some of these do actually have performance impacts when compiling with optimization level `-O3`.

File: src/driverstation/include/ssh_ui.hpp [hunk 2]
```diff
@@ -0,0 +1,771 @@
...
+							case 36: this->current_foreground_color = (this->current_text_effects & BRIGHT) ? Color{  0, 255, 255, 255} : Color{  0, 205, 205, 255}; break;  // Cyan Foreground
+							case 37: this->current_foreground_color = (this->current_text_effects & BRIGHT) ? Color{255, 255, 255, 255} : Color{229, 229, 229, 255}; break;  // White Foreground
+							case 38:
+								switch(parameters[i += 2].value){
+									case 5:  // 8-bit Color Mode
+										{
+											uint8_t n = parameters[i + 2].value;
+
+											// NOTE: The standard palette options modify the `n` parameter such that the next pass actually handles the color application
+											if(n >= 0 && n <= 7){  // Standard foreground color
```
[the remark this answers is not in the document]
Comment (PhocaeaUnitas) [reply src/driverstation/include/ssh_ui.hpp]: See [40be54a](https://github.com/Colorado-School-of-Mines-Robotics-Club/MATE_ROV_Programming/commit/40be54aa21dde2c1cf1b5c6848df74e4cb01a223#diff-e4d02ec2f011f01f0159acd6f4bd5bf98eafde3389acf8a6ceb66a36aeac33a0L604-R604)

File: src/driverstation/include/ssh_ui.hpp [hunk 3]
```diff
@@ -0,0 +1,771 @@
...
+									this->current_text_mode = TextMode::DEFAULT;
+									break;
+								default:
+									this->current_text_mode = TextMode::OPERATING_SYSTEM_COMMAND;
+									break;
+							}
+							break;
+						case TextMode::AWAITING_STRING_TERMINATOR:
+							switch(current_character){
+								case '0x1B':  // Escape, possibly String Terminator
```
[the remark this answers is not in the document]
Comment (PhocaeaUnitas) [reply src/driverstation/include/ssh_ui.hpp]: See [40be54a](https://github.com/Colorado-School-of-Mines-Robotics-Club/MATE_ROV_Programming/commit/40be54aa21dde2c1cf1b5c6848df74e4cb01a223#diff-e4d02ec2f011f01f0159acd6f4bd5bf98eafde3389acf8a6ceb66a36aeac33a0L257-R257)

File: src/driverstation/include/ssh_ui.hpp [hunk 4]
```diff
@@ -0,0 +1,771 @@
...
+								while(this->current_csi_parameters.size() > 0) this->current_csi_parameters.pop_back();
+								this->current_text_mode = TextMode::DEFAULT;
+							}
+							break;
+						case TextMode::OPERATING_SYSTEM_COMMAND:
+							switch(current_character){
+								case '\a':  // For historical reasons, xterm can end OSCs with a bell
+									this->current_text_mode = TextMode::DEFAULT;
+									break;
+								case '0x1B':  // Escape, possibly String Terminator
```
[the remark this answers is not in the document]
Comment (PhocaeaUnitas) [reply src/driverstation/include/ssh_ui.hpp]: See [40be54a](https://github.com/Colorado-School-of-Mines-Robotics-Club/MATE_ROV_Programming/commit/40be54aa21dde2c1cf1b5c6848df74e4cb01a223#diff-e4d02ec2f011f01f0159acd6f4bd5bf98eafde3389acf8a6ceb66a36aeac33a0L240-R240)

File: src/driverstation/include/dynamic_texture.hpp [hunk 5]
```diff
@@ -0,0 +1,395 @@
...
+		// TODO: Are these unloads necessary if raylib handles all unloading when the program exits?
+		void unload() override{
+			UnloadRenderTexture(this->render_texture);
+		}
+
+		Texture2D current_texture() override{
+			return this->render_texture.texture;
+		}
+
+		void on_click(float x, float y) override{}
```
[the remark this answers is not in the document]
Comment (PhocaeaUnitas) [reply src/driverstation/include/dynamic_texture.hpp]: See [40be54a](https://github.com/Colorado-School-of-Mines-Robotics-Club/MATE_ROV_Programming/commit/40be54aa21dde2c1cf1b5c6848df74e4cb01a223#diff-0e6222c0c79b21dc1af73b516d394ec72339ca3798e216ea618bd23100b694ebL62-R62)

File: src/driverstation/include/dynamic_texture.hpp [hunk 6]
```diff
@@ -0,0 +1,395 @@
...
+
+		Texture2D current_texture() override{
+			unsigned int expected_frame_number = (unsigned int)(GetTime() * this->frame_rate) % this->frame_count;
+			// Multiply by 4 for the 4 channels in an RGBA image
+			unsigned int frame_data_offset = this->frame_data.width * this->frame_data.height * 4 * expected_frame_number;
+			UpdateTexture(this->texture, (unsigned char*)this->frame_data.data + frame_data_offset);
+			return this->texture;
+		}
+
+		void on_click(float x, float y) override{}
```
[the remark this answers is not in the document]
Comment (PhocaeaUnitas) [reply src/driverstation/include/dynamic_texture.hpp]: See [40be54a](https://github.com/Colorado-School-of-Mines-Robotics-Club/MATE_ROV_Programming/commit/40be54aa21dde2c1cf1b5c6848df74e4cb01a223#diff-0e6222c0c79b21dc1af73b516d394ec72339ca3798e216ea618bd23100b694ebL200-R200)

File: src/driverstation/include/estop_ui.hpp [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+				// Rectangle this_rectangle = this->rectangle();
+				// for(uint8_t _ = 0; _ < mollyguard_level; _++){
+				// 	DrawRectangleRec(this_rectangle, ColorAlpha(LIGHTGRAY, .25f));
+				// }
+			}
+			
+			return dynamic_texture::DynamicRenderTexture2D::current_texture();
+		}
+
+		void on_click(float x, float y) override{
```
[the remark this answers is not in the document]
Comment (PhocaeaUnitas) [reply src/driverstation/include/estop_ui.hpp]: See [40be54a](https://github.com/Colorado-School-of-Mines-Robotics-Club/MATE_ROV_Programming/commit/40be54aa21dde2c1cf1b5c6848df74e4cb01a223#diff-e8a70cdef6d2873c987c607e6e340a3f6437fe355da6f145cd332d9797048e3eL32-R32)

File: src/driverstation/include/ssh_interface.hpp [hunk 8]
```diff
@@ -0,0 +1,235 @@
...
+		            ssh_clean_pubkey_hash(&hash);
+
+		            return -1;
+		        case SSH_KNOWN_HOSTS_NOT_FOUND:
+		            fprintf(stderr, "Could not find known host file.\n");
+		            fprintf(stderr, "If you accept the host key here, the file will be"
+		                    "automatically created.\n");
+
+		            /* FALL THROUGH to SSH_SERVER_NOT_KNOWN behavior */
+
```
[the remark this answers is not in the document]
Comment (PhocaeaUnitas) [reply src/driverstation/include/ssh_interface.hpp]: See [40be54a](https://github.com/Colorado-School-of-Mines-Robotics-Club/MATE_ROV_Programming/commit/40be54aa21dde2c1cf1b5c6848df74e4cb01a223#diff-21afed6ef9ec3a758c8bec0a5380ca1cf5cd5937d330b53d0cefa4efa694bdf7L71-R71)

File: src/driverstation/include/ssh_ui.hpp [hunk 9]
```diff
@@ -0,0 +1,771 @@
...
+					break;
+				case 'X':  // Erase Characters
+					{
+						// TODO: Known issues when deleting with wraparound text, make better
+						// Could steganography or similar be used to record when lines wrap vs. newline?
+						int characters_to_delete = 1;
+						if(parameters.size() > 0) characters_to_delete = nonZeroOr(parameters[0].value, 1);
+
+						float width = this->history_rectangle.width - (characters_to_delete * this->character_spacing.x);
+						Texture2D texture = this->current_texture();
```
[the remark this answers is not in the document]
Comment (PhocaeaUnitas) [reply src/driverstation/include/ssh_ui.hpp]: See [40be54a](https://github.com/Colorado-School-of-Mines-Robotics-Club/MATE_ROV_Programming/commit/40be54aa21dde2c1cf1b5c6848df74e4cb01a223#diff-e4d02ec2f011f01f0159acd6f4bd5bf98eafde3389acf8a6ceb66a36aeac33a0L553-L554)

File: src/driverstation/include/ssh_ui.hpp [hunk 10]
```diff
@@ -0,0 +1,771 @@
...
+				case 'l':  // Disable bracketed paste - Ignore
+					break;
+				// TODO: Get working?
+				case 'r':  // Set Scrolling Region - Ignore
+					break;
+				case 't':  // Window Manipulation - Ignore
+					break;
+				case 'm':  // Select Graphic Rendition parameters
+					// TODO: Rework to make use of CsiParameter.type property
+					for(int i = 0; i < parameters.size(); i += 2){  // NOTE: += 2 to skip semicolons separating parameters
```
[the remark this answers is not in the document]
Comment (PhocaeaUnitas) [reply src/driverstation/include/ssh_ui.hpp]: See [40be54a](https://github.com/Colorado-School-of-Mines-Robotics-Club/MATE_ROV_Programming/commit/40be54aa21dde2c1cf1b5c6848df74e4cb01a223#diff-e4d02ec2f011f01f0159acd6f4bd5bf98eafde3389acf8a6ceb66a36aeac33a0L581-R581)

File: src/driverstation/include/ssh_ui.hpp (hunk 1 above)
Comment (PhocaeaUnitas) [reply src/driverstation/include/ssh_ui.hpp]: See [40be54a](https://github.com/Colorado-School-of-Mines-Robotics-Club/MATE_ROV_Programming/commit/40be54aa21dde2c1cf1b5c6848df74e4cb01a223#diff-e4d02ec2f011f01f0159acd6f4bd5bf98eafde3389acf8a6ceb66a36aeac33a0L710-R710)

File: src/driverstation/include/driverstation_ui.hpp [hunk 11]
```diff
@@ -0,0 +1,156 @@
...
+					DrawFPS(1800, 20);
+				}
+
+				return !WindowShouldClose();
+			}
+
+		protected:
+			std::unique_ptr<dynamic_texture::MultiCellDynamicTextureGrid2D<12, 12>> multi_cell;
+			std::shared_ptr<gui::SshDisplay> ssh_display;
+			SshConnectPopupMode ssh_connect_popup_mode;
```
[the remark this answers is not in the document]
Comment (PhocaeaUnitas) [reply src/driverstation/include/driverstation_ui.hpp:148]: Use C++20 and I'll use modules instead

**Comment by yameatmeyourdead** [comment]
Closing this, this branch has been replaced by driverstation-rosified. Will not merge.

**Pull request closed without merging**